openshift-kube-apiserver/openshiftkubeapiserver/sdn_readyz_wait.go

```diff
@@ -78,6 +83,31 @@ func (c *aggregatedAPIServiceAvailabilityCheck) checkForConnection(context gener
 		panic(err)
 	}
 
+	ctx := gocontext.TODO()
+
+	// if the kubernetes.default.svc needs an endpoint and this is the only apiserver than can fulfill it, then we don't
+	// wait for reachability. We wait for other conditions, but unreachable apiservers correctly 503 for clients.
+	kubeEndpoints, err := kubeClient.CoreV1().Endpoints("default").Get(ctx, "kubernetes", metav1.GetOptions{})
+	switch {
+	case apierrors.IsNotFound(err):
+		utilruntime.HandleError(fmt.Errorf("%s did not find a kubernetes.default.svc endpoint", c.Name()))
+		return
+	case err != nil:
+		utilruntime.HandleError(fmt.Errorf("%s unable to read a kubernetes.default.svc endpoint: %w", c.Name(), err))
```

Comment:
I think this will be the case when the check hasn't passed yet and the server is not ready.

Reply:
> I think this will be the case when the check hasn't passed yet and the server is not ready.

It's a self-connection so readiness doesn't matter for self-connections

> I was just wondering how frequent this case might be.
If we are unlucky then we might end up here before etcd "becomes ready".

it's possible that's happens, but in such a case the etcd readiness check will prevent our readiness and while not fully correct I think this is the best balance, leading your very important question:

> Now I'm wondering why we added the readiness checks in the first place.

We want to make our best effort to avoid having an instance marked ready and getting traffic to the aggregated openshift/oauth apiservers and returning a 503.  It's not catastrophic to get this wrong and return 503s, but this check makes our master reboot scenario much more consistent.  Prior to this check, we raced with the CNI to make the aggregated apiservers routable from the host before the kube-apiserver went ready.  On upgrades the MCO restarts every node, which made this scenario quite common. But at the time I developed this, we only had HA, not single node or microshift.

Reply:
> it's possible that's happens, but in such a case the etcd readiness check will prevent our readiness and while not fully correct I think this is the best balance, leading your very important question:

I think that in such a case unready etcd will enable our readiness because `GET`ting endpoints will fail with an error leading to closing the `c.close` channel (https://github.com/openshift/kubernetes/blob/aebb7e12f55e358517e8467586e756e0d87c3f9d/openshift-kube-apiserver/openshiftkubeapiserver/sdn_readyz_wait.go#L76).

If this is a common scenario then we will never truly evaluate readiness of aggregated APIs resulting in returning 503 and racing with the CNI to make the aggregated apiservers routable. 

Am I right?

Reply:
ok, discussed in Slack. It looks like our main use-case is `SNO` and on `HA` it is a best-effort.